### PR TITLE
[rfc] generic + returnable dynamic outputs

### PR DESCRIPTION
**Pull request by dpeng817**
Allows DynamicOutput to be used as a generic return type on ops. When using DynamicOut, a list of dynamic outputs is the expected return. Annotating as just a single DynamicOut will result in an error.

### CODE REVIEW

**Review by alangenfeld — CHANGES_REQUESTED** [review]
this seems legit to me, should we update docs ?

to you for BK failing and one inline

File: python_modules/dagster/dagster/core/types/dagster_type.py [hunk 1]
```diff
@@ -877,6 +881,30 @@ def resolve_dagster_type(dagster_type: object) -> DagsterType:
...
+    )
+
+    check.invariant(
+        not (isinstance(dagster_type, type) and issubclass(dagster_type, DagsterType)),
+        "Do not pass runtime type classes. Got {}".format(dagster_type),
+    )
+
+    if dagster_type == DynamicOutput or get_origin(dagster_type) == DynamicOutput:
+        raise DagsterInvariantViolationError(
+            "Op annotated with return type DynamicOutput. DynamicOutputs can only be returned in the context of a List or Mapping. If only one output is needed, use the Output API."
```
[the remark this answers is not in the document]
Comment (dpeng817) [reply python_modules/dagster/dagster/core/types/dagster_type.py:899]: Aspirational about being able to provide a dictionary of keys to dynamicoutputs, but meh

**Comment by dpeng817** [comment]
cc @alangenfeld will follow up with docs in another PR, since we also need to do docs for the regular generic outputs.

File: python_modules/dagster/dagster_tests/core_tests/definitions_tests/decorators_tests/test_op.py [hunk 2]
```diff
@@ -900,3 +900,170 @@ def the_op() -> Tuple[Output[int], Output[str]]:
...
+    result = execute_op_in_graph(basic)
+    assert result.success
+
+    # Equivalent behavior in the dynamic yield case
+    @op(out=DynamicOut(dagster_type=str, is_required=False))
+    def basic_yield():
+        pass
+
+    result = execute_op_in_graph(basic_yield)
+    assert result.success
```
Comment (alangenfeld) [inline python_modules/dagster/dagster_tests/core_tests/definitions_tests/decorators_tests/test_op.py:1054]: maybe leave a comment linking to that issue that this behavior isnt exactly what we want yet

**Review by alangenfeld — APPROVED** [review]

**Pull request merged**